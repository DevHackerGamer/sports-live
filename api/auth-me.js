const { getUserType } = require('../lib/auth');

// Unified handler: resolves user type via Clerk when available, else allows dev header override
module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-User-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'GET') return res.status(405).json({ success: false, error: 'Method not allowed' });

  try {
    const type = (await getUserType(req)) || 'user';
    const isAdmin = String(type).toLowerCase() === 'admin';
    return res.status(200).json({ success: true, type, isAdmin });
  } catch (error) {
    console.error('auth-me error:', error);
    return res.status(200).json({ success: true, type: 'user', isAdmin: false });
  }
};
