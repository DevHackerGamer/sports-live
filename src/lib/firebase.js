// Firebase client initialization (App + Realtime Database)
import { initializeApp, getApps, getApp } from 'firebase/app';
import { getAnalytics } from 'firebase/analytics';
import {
  getDatabase,
  ref as dbRef,
  update as dbUpdate,
  set as dbSet,
  child as dbChild,
  get as dbGet,
  onValue as dbOnValue  // <-- add this
} from 'firebase/database';

// Firebase web config (public by design)
const firebaseConfig = {
  apiKey: 'AIzaSyDzZ0CAodwCYHgQJahsgcxJrmWiIIbVGM',
  authDomain: 'sports-live-c25ca.firebaseapp.com',
  databaseURL: 'https://sports-live-c25ca-default-rtdb.firebaseio.com',
  projectId: 'sports-live-c25ca',
  storageBucket: 'sports-live-c25ca.firebasestorage.app',
  messagingSenderId: '532399482994',
  appId: '1:532399482994:web:3d1353393886b64dc1baeb',
  measurementId: 'G-8JT6X5DF6B',
};

// Singleton app instance across HMR/dev reloads
const app = getApps().length ? getApp() : initializeApp(firebaseConfig);

// Realtime Database instance
export const db = getDatabase(app);

// Modular functions for convenience
export const ref = dbRef;
export const update = dbUpdate;
export const set = dbSet;
export const child = dbChild;
export const get = dbGet;
export const onValue = dbOnValue;  // <-- export onValue

// Guard analytics for browser-only environments
export const initAnalytics = () => {
  if (typeof window !== 'undefined' && 'measurementId' in firebaseConfig) {
    try {
      return getAnalytics(app);
    } catch (_) {
      // ignore if analytics not supported
    }
  }
  return null;
};

export default app;
