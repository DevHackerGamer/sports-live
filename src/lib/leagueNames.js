// Centralized league name mapping for ESPN league codes
// This provides consistent display names across the entire application

export const LEAGUE_NAMES = {
  // ESPN league codes
  'eng.1': 'Premier League',
  'esp.1': 'La Liga',
  'ita.1': 'Serie A',
  'ger.1': 'Bundesliga',
  'fra.1': 'Ligue 1',
  'uefa.champions': 'Champions League',
  'ned.1': 'Eredivisie',
  'por.1': 'Primeira Liga',
  'eng.2': 'Championship',
  'bra.1': 'Campeonato Brasileiro SÃ©rie A',
  'conmebol.libertadores': 'Copa Libertadores',
  
  // Football-data.org codes (legacy)
  'PL': 'Premier League',
  'PD': 'La Liga',
  'SA': 'Serie A',
  'BL1': 'Bundesliga',
  'FL1': 'Ligue 1',
  'CL': 'Champions League',
  'DED': 'Eredivisie',
  'PPL': 'Primeira Liga',
  'ELC': 'Championship'
};

/**
 * Get display name for a league code
 * @param {string} code - ESPN league code (e.g., 'eng.1') or name with bracket notation (e.g., 'Bundesliga [ger.1]')
 * @returns {string} Display name (e.g., 'Premier League')
 */
export function getLeagueName(code) {
  if (!code) return 'Unknown League';
  
  // Handle strings with bracket notation like "Bundesliga [ger.1]"
  const bracketMatch = /^(.*?)\s*\[([^\]]+)\]$/.exec(String(code));
  if (bracketMatch) {
    const [, namePart, codePart] = bracketMatch;
    // Try to get the name from the code part first
    const nameFromCode = LEAGUE_NAMES[codePart];
    if (nameFromCode) return nameFromCode;
    // Otherwise return the name part without brackets
    return namePart.trim();
  }
  
  return LEAGUE_NAMES[code] || code;
}

/**
 * Get league code from display name or return as-is if already a code
 * @param {string} nameOrCode - League name or code
 * @returns {string} League code
 */
export function getLeagueCode(nameOrCode) {
  if (!nameOrCode) return '';
  
  // If it's already a code, return it
  if (LEAGUE_NAMES[nameOrCode]) return nameOrCode;
  
  // Find code by name
  const entry = Object.entries(LEAGUE_NAMES).find(([_, name]) => name === nameOrCode);
  return entry ? entry[0] : nameOrCode;
}

/**
 * Format competition object to ensure it has proper display name
 * @param {Object} competition - Competition object with name/code
 * @returns {Object} Competition with displayName
 */
export function formatCompetition(competition) {
  if (!competition) return { name: 'Unknown League', displayName: 'Unknown League' };
  
  const name = competition.name || competition.code || competition.id || '';
  const displayName = getLeagueName(name);
  
  return {
    ...competition,
    displayName,
    code: competition.code || name
  };
}
