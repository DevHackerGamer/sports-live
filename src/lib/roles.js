// Client-side role utilities to interpret Clerk metadata consistently

export function getUserRoles(user) {
  if (!user) return [];
  const raw = [
    user?.privateMetadata?.type,
    user?.publicMetadata?.type,

  ]
    .concat(Array.isArray(user?.privateMetadata?.roles) ? user.privateMetadata.roles : [])
    .concat(Array.isArray(user?.publicMetadata?.roles) ? user.publicMetadata.roles : [])
    .filter(Boolean)
    .map(v => String(v).toLowerCase());
  return Array.from(new Set(raw));
}

export function isAdminFromUser(user) {
  const roles = getUserRoles(user);
  return roles.includes('admin');
}
