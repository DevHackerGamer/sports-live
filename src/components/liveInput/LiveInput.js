import React, { useState, useEffect, useRef } from 'react';
import { useUser } from '@clerk/clerk-react';
import { isAdminFromUser } from '../../lib/roles';
import '../../styles/LiveInput.css';

// Admin-only screen for entering live match data.
// Accepts optional isAdmin prop from parent; falls back to Clerk metadata if not provided.
const LiveInput = ({ isAdmin: isAdminProp }) => {
  const { user } = useUser();

  // All hooks must be declared unconditionally and in the same order on every render.
  const [isPaused, setIsPaused] = useState(true);
  const [matchTime, setMatchTime] = useState(0); // time in seconds
  const [score, setScore] = useState({ home: 0, away: 0 });
  const [possession, setPossession] = useState(50);
  const [events, setEvents] = useState([]);
  const [newEvent, setNewEvent] = useState({
    type: 'goal',
    team: 'home',
    player: '',
    minute: '0',
    description: ''
  });
  const timerRef = useRef(null);

  // Derive admin after hooks are declared to avoid conditional hooks.
  const isAdmin = typeof isAdminProp === 'boolean' ? isAdminProp : isAdminFromUser(user);

  // Format seconds to MM:SS
  const formatTime = (seconds) => {
    const mins = Math.floor(seconds / 60);
    const secs = seconds % 60;
    return `${mins.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
  };

  // Update the minute field when matchTime changes
  useEffect(() => {
    const minutes = Math.floor(matchTime / 60);
    setNewEvent(prev => ({...prev, minute: minutes.toString()}));
  }, [matchTime]);

  // Start/stop the timer since its a live match input
  useEffect(() => {
    if (!isPaused) {
      timerRef.current = setInterval(() => {
        setMatchTime(prev => prev + 1);
      }, 1000);
    } else {
      clearInterval(timerRef.current);
    }

    return () => clearInterval(timerRef.current);
  }, [isPaused]);

  const handleEventAdd = () => {
    if (newEvent.player) {
      const event = {
        id: Date.now(),
        type: newEvent.type,
        team: newEvent.team,
        player: newEvent.player,
        minute: Math.floor(matchTime / 60), // Always use current match time
        description: newEvent.description,
        timestamp: new Date()
      };
      
      //prev : setEvents([...events, event]);
      setEvents([...events, { ...event, id: Date.now() }]); //now to update local state
      
      // Update score if it's a goal
      if (newEvent.type === 'goal') {
        setScore(prev => ({
          ...prev,
          [newEvent.team]: prev[newEvent.team] + 1
        }));
      }
      
      // Reset form but keep type and team
      setNewEvent({
        ...newEvent,
        player: '',
        description: ''
      });
    }
  };

  const removeEvent = (id) => {
    setEvents(events.filter(event => event.id !== id));
  };

  const togglePause = () => {
    setIsPaused(!isPaused);
  };

  const resetTimer = () => {
    setIsPaused(true);
    setMatchTime(0);
  };

  if (!isAdmin) {
    return (
      <div className="live-input">
        <h2>Live Match Input</h2>
        <p style={{ color: '#b00' }}>Access denied: Admin role required.</p>
      </div>
    );
  }

  return (
    <div className="live-input">
      <h2>Live Match Input</h2>
      
      <div className="match-controls">
        <div className="time-control">
          <label>Match Time:</label>
          <div className="time-display">{formatTime(matchTime)}</div>
          <div className="timer-controls">
            <button className="timer-btn start-btn" onClick={togglePause}>
              {isPaused ? 'Start' : 'Pause'}
            </button>
            <button className="timer-btn reset-btn" onClick={resetTimer}>
              Reset
            </button>
          </div>
        </div>
        
        <div className="score-control">
          <label>Score:</label>
          <div className="score-inputs">
            <input 
              type="number" 
              min="0" 
              value={score.home} 
              onChange={(e) => setScore({...score, home: parseInt(e.target.value) || 0})}
            />
            <span>:</span>
            <input 
              type="number" 
              min="0" 
              value={score.away} 
              onChange={(e) => setScore({...score, away: parseInt(e.target.value) || 0})}
            />
          </div>
        </div>
        
        <div className="possession-control">
          <label>Possession (%):</label>
          <div className="possession-slider">
            <input 
              type="range" 
              min="0" 
              max="100" 
              value={possession} 
              onChange={(e) => setPossession(parseInt(e.target.value))}
            />
            <div className="possession-values">
              <span>Home: {possession}%</span>
              <span>Away: {100 - possession}%</span>
            </div>
          </div>
        </div>
      </div>
      
      <div className="event-input">
        <h3>Add Match Event</h3>
        <div className="event-form">
          <div className="form-row">
            <select 
              value={newEvent.type}
              onChange={(e) => setNewEvent({...newEvent, type: e.target.value})}
            >
              <option value="goal">Goal</option>
              <option value="yellow">Yellow Card</option>
              <option value="red">Red Card</option>
              <option value="substitution">Substitution</option>
              <option value="injury">Injury</option>
              <option value="yellowred">Yellow-Red Card</option>
              <option value="penalty">Penalty</option>
              <option value="owngoal">Own Goal</option>
              <option value="foul">Foul</option>
              <option value="corner">Corner</option>
              <option value="freekick">Free Kick</option>
              <option value="offside">Offside</option>
              <option value="save">Save</option>
              <option value="other">Other</option>
            </select>
            
            <select 
              value={newEvent.team}
              onChange={(e) => setNewEvent({...newEvent, team: e.target.value})}
            >
              <option value="home">Home Team</option>
              <option value="away">Away Team</option>
            </select>
            
            <div className="minute-display">
              Minute: {Math.floor(matchTime / 60)}
            </div>
          </div>
          
          <div className="form-row">
            <input 
              type="text" 
              placeholder="Player Name" 
              value={newEvent.player}
              onChange={(e) => setNewEvent({...newEvent, player: e.target.value})}
            />
          </div>
          
          <textarea 
            placeholder="Event description" 
            value={newEvent.description}
            onChange={(e) => setNewEvent({...newEvent, description: e.target.value})}
          />
          
          <button onClick={handleEventAdd}>Add Event</button>
        </div>
      </div>
      
      <div className="events-log">
        <h3>Events Timeline</h3>
        {events.length === 0 ? (
          <p className="no-events">No events recorded yet.</p>
        ) : (
          events.map(event => (
            <div key={event.id} className={`event-log-item ${event.type}`}>
              <span className="event-time">{event.minute}'</span>
              <span className="event-type">
                {event.type === 'goal' && '⚽'}
                {event.type === 'yellow' && '🟨'}
                {event.type === 'red' && '🟥'}
                {event.type === 'substitution' && '🔁'}
                {event.type === 'injury' && '🤕'}
                {event.type === 'yellowred' && '🟨🟥'}
                {event.type === 'penalty' && '⚽ (P)'}
                {event.type === 'owngoal' && '⚽ (OG)'}
                {event.type === 'foul' && '🚫'}
                {event.type === 'corner' && '� corner'}
                {event.type === 'freekick' && '🎯'}
                {event.type === 'offside' && '🚩'}
                {event.type === 'save' && '🧤 '}
                {event.type === 'other' && '🔔'}
              </span>
              <span className="event-details">
                {event.team === 'home' ? 'Home' : 'Away'} - {event.player}: {event.description}
              </span>
              <button className="remove-btn" onClick={() => removeEvent(event.id)}>×</button>
            </div>
          ))
        )}
      </div>
    </div>
  );
};

export default LiveInput;