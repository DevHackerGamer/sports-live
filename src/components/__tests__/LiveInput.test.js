import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import LiveInput from '../liveInput/LiveInput';
import * as ClerkReact from '@clerk/clerk-react';
import * as roles from '../../lib/roles';
import * as api from '../../lib/api';

// Mock Clerk
jest.mock('@clerk/clerk-react', () => ({
  useUser: jest.fn(),
}));

// Mock API client
jest.mock('../../lib/api', () => ({
  apiClient: {
    getMatch: jest.fn(),
    getTeams: jest.fn(),
    addMatchEvent: jest.fn(),
    deleteMatchEvent: jest.fn(),
    updateMatch: jest.fn(),
    getMatchStatistics: jest.fn(),
    updateMatchStatistics: jest.fn(),
  },
}));

// Mock child components
jest.mock('../liveInput/LineupsAdminModal', () => ({ match, onClose }) => (
  <div data-testid="lineups-admin-modal">
    Lineups Admin Modal for {match?.id}
    <button onClick={onClose}>Close Lineups</button>
  </div>
));

jest.mock('../liveInput/CommentaryAdminModal', () => ({ matchId, isOpen, onClose }) => (
  <div data-testid="commentary-admin-modal">
    Commentary Admin Modal for match {matchId}
    <button onClick={onClose}>Close Commentary</button>
  </div>
));

describe('LiveInput Component', () => {
  const mockOnBackToMatch = jest.fn();
  const mockUser = { id: 'user123', firstName: 'John' };
  const mockAdminUser = { 
    id: 'admin123', 
    firstName: 'Admin',
    privateMetadata: { type: 'admin' }
  };

  const mockMatch = {
    id: 'match1',
    homeTeam: { name: 'Arsenal', crest: '/arsenal.png' },
    awayTeam: { name: 'Chelsea', crest: '/chelsea.png' },
    status: 'IN_PLAY',
    createdByAdmin: true,
    clock: {
      running: true,
      elapsed: 1200, // 20 minutes
      startedAt: new Date(Date.now() - 1200000).toISOString(),
    },
    events: [
      {
        id: 'event1',
        type: 'goal',
        minute: 15,
        team: 'home',
        player: 'Bukayo Saka',
        description: 'Goal - Arsenal - Bukayo Saka',
      },
    ],
  };

  const mockTeamsData = {
    data: [
      { _id: 'team1', name: 'Arsenal', crest: '/arsenal.png' },
      { _id: 'team2', name: 'Chelsea', crest: '/chelsea.png' },
    ],
  };

  const mockPlayersData = {
    players: [
      { name: 'Bukayo Saka', position: 'Forward' },
      { name: 'Martin Ødegaard', position: 'Midfielder' },
      { name: 'Reece James', position: 'Defender' },
      { name: 'Raheem Sterling', position: 'Forward' },
    ],
  };

  const mockStatsData = {
    possession: { home: 60, away: 40 },
    shotsOnTarget: { home: 5, away: 3 },
    shotsOffTarget: { home: 8, away: 6 },
    totalShots: { home: 13, away: 9 },
    corners: { home: 4, away: 2 },
    fouls: { home: 10, away: 12 },
    yellowCards: { home: 1, away: 2 },
    redCards: { home: 0, away: 0 },
    offsides: { home: 2, away: 1 },
    saves: { home: 2, away: 4 },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
    
    // Default mocks
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(true);
    
    // Mock API responses
    api.apiClient.getMatch.mockResolvedValue({ data: mockMatch });
    api.apiClient.getTeams.mockResolvedValue(mockTeamsData);
    api.apiClient.addMatchEvent.mockResolvedValue({ data: {} });
    api.apiClient.deleteMatchEvent.mockResolvedValue({});
    api.apiClient.updateMatch.mockResolvedValue({});
    api.apiClient.getMatchStatistics.mockResolvedValue(mockStatsData);
    api.apiClient.updateMatchStatistics.mockResolvedValue(mockStatsData);
    
    // Mock fetch for players
    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve(mockPlayersData),
    });
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders access denied for non-admin users', () => {
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(false);

    render(<LiveInput match={mockMatch} onBackToMatch={mockOnBackToMatch} />);

    expect(screen.getByText('Access denied: Admin role required.')).toBeInTheDocument();
  });

  it('renders no match selected message', () => {
    render(<LiveInput isAdmin={true} onBackToMatch={mockOnBackToMatch} />);

    expect(screen.getByText('No match selected. Select a match first from Live Sports.')).toBeInTheDocument();
  });

  it('renders live input interface for admin with match', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    expect(screen.getByText('Live Match Input')).toBeInTheDocument();
    expect(screen.getByText('Arsenal')).toBeInTheDocument();
    expect(screen.getByText('Chelsea')).toBeInTheDocument();
    expect(screen.getByText('↩ Back')).toBeInTheDocument();
  });

  it('displays match timer and controls', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    expect(screen.getByText('20:00')).toBeInTheDocument(); // 20 minutes in MM:SS format
    expect(screen.getByText('⏸')).toBeInTheDocument(); // Pause button
  });

  it('starts and pauses timer', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const pauseButton = screen.getByText('⏸');
    fireEvent.click(pauseButton);

    expect(screen.getByText('▶')).toBeInTheDocument(); // Now shows play button

    // Advance time and verify timer doesn't run when paused
    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(screen.getByText('20:00')).toBeInTheDocument(); // Time should not change
  });

  it('adds new match event', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    // Fill event form
    const playerInput = screen.getByPlaceholderText('Player name');
    const descriptionInput = screen.getByPlaceholderText('Event description');
    const addButton = screen.getByText('Add Event');

    fireEvent.change(playerInput, { target: { value: 'Martin Ødegaard' } });
    fireEvent.change(descriptionInput, { target: { value: 'Great goal from outside the box' } });

    await act(async () => {
      fireEvent.click(addButton);
    });

    expect(api.apiClient.addMatchEvent).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        type: 'goal',
        player: 'Martin Ødegaard',
        description: 'Great goal from outside the box',
      }),
      { userType: 'admin' }
    );
  });

  it('adds substitution event', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    // Change event type to substitution
    const eventTypeSelect = screen.getByDisplayValue('Goal');
    fireEvent.change(eventTypeSelect, { target: { value: 'substitution' } });

    // Fill substitution fields
    const playerOutInput = screen.getByPlaceholderText('Player OUT');
    const playerInInput = screen.getByPlaceholderText('Player IN');
    const addButton = screen.getByText('Add Event');

    fireEvent.change(playerOutInput, { target: { value: 'Bukayo Saka' } });
    fireEvent.change(playerInInput, { target: { value: 'Martin Ødegaard' } });

    await act(async () => {
      fireEvent.click(addButton);
    });

    expect(api.apiClient.addMatchEvent).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        type: 'substitution',
        playerOut: 'Bukayo Saka',
        playerIn: 'Martin Ødegaard',
      }),
      { userType: 'admin' }
    );
  });

  it('removes event from timeline', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const removeButtons = screen.getAllByText('×');
    await act(async () => {
      fireEvent.click(removeButtons[0]);
    });

    expect(api.apiClient.deleteMatchEvent).toHaveBeenCalledWith(
      'match1',
      'event1',
      { userType: 'admin' }
    );
  });

  it('handles half time button click', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const halfTimeButton = screen.getByText('⏱ Half Time');
    await act(async () => {
      fireEvent.click(halfTimeButton);
    });

    expect(api.apiClient.addMatchEvent).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        type: 'half_time',
        description: 'Half Time',
      }),
      { userType: 'admin' }
    );
  });

  it('handles full time button click', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const fullTimeButton = screen.getByText('🏁 Full Time');
    await act(async () => {
      fireEvent.click(fullTimeButton);
    });

    expect(api.apiClient.addMatchEvent).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        type: 'match_end',
        description: 'Full Time',
      }),
      { userType: 'admin' }
    );

    expect(api.apiClient.updateMatch).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        status: 'FINISHED',
      }),
      { userType: 'admin' }
    );
  });

  it('updates possession when clicked on possession bar', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const possessionBar = screen.getByTitle(/Click to set possession/);
    
    // Mock getBoundingClientRect for the possession bar
    possessionBar.getBoundingClientRect = jest.fn(() => ({
      left: 0,
      width: 200,
      top: 0,
      height: 20,
      right: 200,
      bottom: 20,
    }));

    fireEvent.click(possessionBar, { clientX: 100 }); // Click at 50%

    expect(screen.getByText('50%')).toBeInTheDocument();
    expect(screen.getByText('50%')).toBeInTheDocument(); // Both teams should show 50%
  });

  it('shows and hides statistics panel', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const statsButton = screen.getByText('📊 Show Stats');
    fireEvent.click(statsButton);

    expect(screen.getByText('Match Statistics')).toBeInTheDocument();
    expect(screen.getByText('Hide Stats')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Hide Stats'));
    expect(screen.queryByText('Match Statistics')).not.toBeInTheDocument();
  });

  it('updates match statistics', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    // Show stats panel
    fireEvent.click(screen.getByText('📊 Show Stats'));

    // Find and update a statistic
    const shotsInput = screen.getAllByDisplayValue('5')[0]; // Home shots on target
    fireEvent.change(shotsInput, { target: { value: '6' } });

    expect(api.apiClient.updateMatchStatistics).toHaveBeenCalledWith(
      'match1',
      expect.objectContaining({
        shotsOnTarget: { home: 6, away: 3 },
      })
    );
  });

  it('shows and hides lineups modal', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const lineupsButton = screen.getByText('👥 Edit Lineups');
    fireEvent.click(lineupsButton);

    expect(screen.getByTestId('lineups-admin-modal')).toBeInTheDocument();
    expect(screen.getByText('Close Lineups')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Close Lineups'));
    expect(screen.queryByTestId('lineups-admin-modal')).not.toBeInTheDocument();
  });

  it('shows and hides commentary modal', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const commentaryButton = screen.getByText('💬 Live Commentary');
    fireEvent.click(commentaryButton);

    expect(screen.getByTestId('commentary-admin-modal')).toBeInTheDocument();
    expect(screen.getByText('Close Commentary')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Close Commentary'));
    expect(screen.queryByTestId('commentary-admin-modal')).not.toBeInTheDocument();
  });

  it('handles back button click', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const backButton = screen.getByText('↩ Back');
    fireEvent.click(backButton);

    expect(mockOnBackToMatch).toHaveBeenCalled();
  });

  it('displays events timeline', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    expect(screen.getByText('Events Timeline')).toBeInTheDocument();
    expect(screen.getByText('Goal - Arsenal - Bukayo Saka')).toBeInTheDocument();
  });

  it('shows no events message when timeline is empty', async () => {
    const matchWithoutEvents = {
      ...mockMatch,
      events: [],
    };

    await act(async () => {
      render(<LiveInput isAdmin={true} match={matchWithoutEvents} onBackToMatch={mockOnBackToMatch} />);
    });

    expect(screen.getByText('No events recorded yet.')).toBeInTheDocument();
  });

  it('handles API errors gracefully', async () => {
    api.apiClient.addMatchEvent.mockRejectedValue(new Error('Failed to add event'));

    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const playerInput = screen.getByPlaceholderText('Player name');
    const addButton = screen.getByText('Add Event');

    fireEvent.change(playerInput, { target: { value: 'Test Player' } });

    await act(async () => {
      fireEvent.click(addButton);
    });

    // Should not crash, error should be handled
    expect(screen.getByText('Live Match Input')).toBeInTheDocument();
  });

  it('prevents adding events to finished match', async () => {
    const finishedMatch = {
      ...mockMatch,
      status: 'FINISHED',
    };

    await act(async () => {
      render(<LiveInput isAdmin={true} match={finishedMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const addButton = screen.getByText('Add Event');
    expect(addButton).toBeDisabled();
    expect(addButton).toHaveAttribute('title', 'Cannot add events to finished match');
  });

  it('disables half time and full time buttons for finished match', async () => {
    const finishedMatch = {
      ...mockMatch,
      status: 'FINISHED',
    };

    await act(async () => {
      render(<LiveInput isAdmin={true} match={finishedMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    const halfTimeButton = screen.getByText('⏱ Half Time');
    const fullTimeButton = screen.getByText('🏁 Full Time');

    expect(halfTimeButton).toBeDisabled();
    expect(fullTimeButton).toBeDisabled();
  });

  it('formats time correctly', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    // Should format 1200 seconds as 20:00
    expect(screen.getByText('20:00')).toBeInTheDocument();
  });

  it('loads match statistics on mount', async () => {
    await act(async () => {
      render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    });

    expect(api.apiClient.getMatchStatistics).toHaveBeenCalledWith('match1');
  });

  it('matches snapshot with admin and match', async () => {
    await act(async () => {
      const { container } = render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('matches snapshot with non-admin user', () => {
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(false);

    const { container } = render(<LiveInput match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot without match', () => {
    const { container } = render(<LiveInput isAdmin={true} onBackToMatch={mockOnBackToMatch} />);
    expect(container).toMatchSnapshot();
  });

  describe('Timer functionality', () => {
    it('increments timer when running', async () => {
      await act(async () => {
        render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
      });

      // Timer should be running initially
      expect(screen.getByText('20:00')).toBeInTheDocument();

      // Advance time by 1 second
      act(() => {
        jest.advanceTimersByTime(1000);
      });

      expect(screen.getByText('20:01')).toBeInTheDocument();
    });

    it('stops timer when match is finished', async () => {
      const finishedMatch = {
        ...mockMatch,
        status: 'FINISHED',
      };

      await act(async () => {
        render(<LiveInput isAdmin={true} match={finishedMatch} onBackToMatch={mockOnBackToMatch} />);
      });

      // Timer should not increment for finished match
      const initialTime = screen.getByText('20:00').textContent;

      act(() => {
        jest.advanceTimersByTime(5000);
      });

      expect(screen.getByText(initialTime)).toBeInTheDocument();
    });
  });

  describe('Event type handling', () => {
    it('handles different event types correctly', async () => {
      await act(async () => {
        render(<LiveInput isAdmin={true} match={mockMatch} onBackToMatch={mockOnBackToMatch} />);
      });

      const eventTypeSelect = screen.getByDisplayValue('Goal');
      
      // Test yellow card
      fireEvent.change(eventTypeSelect, { target: { value: 'yellow_card' } });
      expect(screen.getByDisplayValue('Yellow Card')).toBeInTheDocument();

      // Test red card
      fireEvent.change(eventTypeSelect, { target: { value: 'red_card' } });
      expect(screen.getByDisplayValue('Red Card')).toBeInTheDocument();

      // Test substitution (should show different inputs)
      fireEvent.change(eventTypeSelect, { target: { value: 'substitution' } });
      expect(screen.getByPlaceholderText('Player OUT')).toBeInTheDocument();
      expect(screen.getByPlaceholderText('Player IN')).toBeInTheDocument();
    });
  });
});