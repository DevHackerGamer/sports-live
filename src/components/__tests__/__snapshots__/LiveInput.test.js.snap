// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveInput Component matches snapshot with admin and match 1`] = `<div />`;

exports[`LiveInput Component matches snapshot with non-admin user 1`] = `
<div>
  <div
    class="live-input"
  >
    <div
      style="display: flex; justify-content: space-between; align-items: center; margin-bottom: 2rem;"
    >
      <h2>
        Live Match Input
      </h2>
      <button
        class="timer-btn"
        style="margin-left: auto;"
      >
        ↩ Back
      </button>
    </div>
    <p
      style="color: rgb(187, 0, 0);"
    >
      Access denied: Admin role required.
    </p>
  </div>
</div>
`;

exports[`LiveInput Component matches snapshot without match 1`] = `
<div>
  <div
    class="live-input"
  >
    <div
      style="display: flex; justify-content: space-between; align-items: center; margin-bottom: 2rem;"
    >
      <h2>
        Live Match Input
      </h2>
      <button
        class="timer-btn"
        style="margin-left: auto;"
      >
        ↩ Back
      </button>
    </div>
    <p>
      No match selected. Select a match first from Live Sports.
    </p>
  </div>
</div>
`;
